design/sd-local.md

```diff
@@ -0,0 +1,164 @@
+## Context
+The build environment on Screwdriver.cd differs from the user execution environment in the following ways:
+- `sd-step` is available
+- Template is available
+- Metadata is available
+- Guarantee build reproducibillity
+
+Therefore, the build results may differ between the build environment on local and the build environment on Screwdriver.cd.
+As a result, User cannot confirm whether the result obtained on CI is the expected result until the build is actually run on Screwdriver.cd.
+
+## Proposal
+
+Provide the function for users to run builds locally.
+
+A simple example is shown below
+
+```
+$ cat screwdriver.yaml
+jobs:
+  test:
+    image: centos7:latest
+    steps:
+      - echo: echo "test"
+$ sdlocal build test # start `test` job in screwdriver.yaml
+echo "test"
+test
+```
+
+## Details
+
+![image](./diagrams/sd-local-flow.puml.png)
+
+
+### Initialization
+
+1. Run `sdlocal build` command
+2. Get JWT from API
+3. Validate screwdriver.yaml
+   - Validate yaml
+   - Get Template/Step information
+4. Pull Launcher Image from Docker Hub
+5. Copy the binaries under `/opt/sd/` from Launcher Container to the common volume
+6. Pull Build Image from Docker Hub
+
+### Run Build
+1. Mount the following and run Build Container
+   - SSH-Agent Socket
+   - The common volume prepared in `Initialization` No.4
+   - Source Directory
+   - Artifacts Directory
+2. Run Local Mode Launcher on Build Container
+   - Run Launcher binary with `--local-mode` option in `run.sh`
+   - Config Build Environment Variables and Environment Variables by env option 
+   - Run Steps got in `Initialization` No.3
+   - Run `sd-cmd`
+     1. Get the binaries from Store
+     2. Run it
+3. Write Artifacts to Artifacts Directory
+4. Output logs to stdout
```

Comment:
can logs also be written to a path under Artifacts directory ?

Reply:
We can output logs under Artifacts directory.
Would you tell me why?

Reply:
console logs might be too long to analyze later. if we persist logs automatically under Artifacts dir, then user can refer back it later, without having to worry about scrolling limitations of terminal.

Reply:
I understand. I think it is better too.